File: codes/golang/341.扁平化嵌套列表迭代器.go

type NestedIterator struct {
	list []NestedInteger
}

func Constructor(nestedList []*NestedInteger) *NestedIterator {
	list := make([]NestedInteger, 0)
	for _, nestedInt := range nestedList {
		list = append(list, *nestedInt)
	}
	return &NestedIterator{list}
}

func (this *NestedIterator) Next() int {
	// hasNext 方法保证了第一个元素一定是整数类型
	res := this.list[0].GetInteger()
	this.list = this.list[1:]
	return res
}

func (this *NestedIterator) HasNext() bool {
	// 循环拆分列表元素，直到列表第一个元素是整数类型
	for len(this.list) > 0 && !this.list[0].IsInteger() {
		// 当列表开头第一个元素是列表类型时，进入循环
		first := this.list[0].GetList()
		this.list = this.list[1:]
		// 将第一个列表打平并按顺序添加到开头
		for i := len(first) - 1; i >= 0; i-- {
			this.list = append([]NestedInteger{first[i]}, this.list...)
		}
	}
	return len(this.list) > 0
}

File: codes/python/341.扁平化嵌套列表迭代器.py


class NestedIterator:
    def __init__(self, nestedList: [NestedInteger]):
        # 不直接用 nestedList 的引用，是因为不能确定它的底层实现
        # 必须保证是 LinkedList，否则下面的 addFirst 会很低效
        self.list = deque(nestedList)

    def next(self) -> int:
        # hasNext 方法保证了第一个元素一定是整数类型
        return self.list.popleft().getInteger()

    def hasNext(self) -> bool:
        # 循环拆分列表元素，直到列表第一个元素是整数类型
        while self.list and not self.list[0].isInteger():
            # 当列表开头第一个元素是列表类型时，进入循环
            first = self.list.popleft().getList()
            # 将第一个列表打平并按顺序添加到开头
            for i in range(len(first)-1, -1, -1):
                self.list.appendleft(first[i])
        return bool(self.list)